platform/engineering/request-for-change/2019-09-20-rfc-cms-based-alerts.md

```diff
@@ -0,0 +1,84 @@
+
+[2] *Technically, a non-engineer comfortable with git and markdown or YAML could update the GitHub-based homepage banner. But those folks are few and far between in the VA content manager ranks.*
+
+
+[3] *PVAMC's legacy VA.gov facility pages are managed via the Teamsite CMS.*
+
+--
+
+## Motivation
+To satisfy this user story:
+
+> *As a VA.gov content manager, I need to reliably publish content updates to VA.gov in less than ten minutes without relying on an engineer in order to deliver time-sensitive information to Veterans.*
+
+This functionality does not currently exist. Each of the alert implementations discussed above has significant limitations in attempting to satisfy the story:
+
+1. **Existing BRD process**. This takes more than ten minutes to run and rebuilds the entire site. Feels very weighty for what may regularly be a small update on a single page.
+1. **Teamsite**. This is going away once legacy VA.gov pages are migrated to WBC VA.gov or retired.
+1. **Sourced from `va.gov-content` on load**. Requires content managers to be comfortable with GitHub and markdown or YAML.
+1. **Sourced from Vets API on load.** No logical connection between the CMS and Vets API exists.
+
+## Possible design approaches
+
+For the first two possibilities, assume that a change to a specific area of the content model causes the CMS to tell Jenkins kick off a specific pipeline with certain parameters. For the third, Jenkins is out of the loop.
+
+### Single page BRD
+
+Jenkins makes a GraphQL query the CMS content API based on given parameters, pulls _only_ the relevant content and flags, builds _only_ the corresponding templates, runs _only_ the relevant tests and checks, and uploads the HTML asset to the existing S3 bucket, overwriting the existing asset. As soon as upload is complete, the page begins being served to users. Alerts are rendered along with the rest of the static content on the page.
+
+|Pros|Cons|
+|----|----|
+|Familiar architecture and process|Somewhat weighty process for minor updates|
+|Can apply existing tests (e.g. link checker, 508)|Inherently slower b/c of tests and checks|
+|No additional client requests needed after page download|
+|Leverages benefits of GraphQL|
+
+### Sourced from S3 bucket on load
+
+Jenkins makes a GraphQL query to the CMS content API based on given parameters, pulls _only_ the relevant content and flags, builds one or more JSON, YAML, or other type of structured data file, and uploads to a dedicated S3 bucket, possibly overwriting the existing asset. On page load, a React component makes an asynchronous call to the S3 bucket, parses the payload for any relevant alert information, and renders the result (if applicable).
+
+|Pros|Cons|
+|----|----|
+|Lightweight and fast|Need to track another kind of asset outside of source control|
+|Leverages benefits of GraphQL|Adds more architectural and process complexity website management|
+| |Requires at least one more client request after page download|
```

Comment:
A HUGE con to me is that this moves business logic into vets-website code.

Reply:
I say that because we also want to make sure that the alert gets published before the call out to send the email alerts.

I imagine other facilities will want to do something similar but probably not the same thing. They may have other email sending providers, want to tweet, etc. Do we need to have a generic "actions" editor for various facilities? And does having this encapsulated somewhere outside of the CMS make it more complicated?